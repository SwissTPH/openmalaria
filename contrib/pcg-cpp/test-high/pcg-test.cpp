/*
 * PCG Random Number Generation for C++
 *
 * Copyright 2014-2017 Melissa O'Neill <oneill@pcg-random.org>,
 *                     and the PCG Project contributors.
 *
 * SPDX-License-Identifier: (Apache-2.0 OR MIT)
 *
 * Licensed under the Apache License, Version 2.0 (provided in
 * LICENSE-APACHE.txt and at http://www.apache.org/licenses/LICENSE-2.0)
 * or under the MIT license (provided in LICENSE-MIT.txt and at
 * http://opensource.org/licenses/MIT), at your option. This file may not
 * be copied, modified, or distributed except according to those terms.
 *
 * Distributed on an "AS IS" BASIS, WITHOUT WARRANTY OF ANY KIND, either
 * express or implied.  See your chosen license for details.
 *
 * For additional information about the PCG random number generation scheme,
 * visit http://www.pcg-random.org/.
 */

/*
 * This file is based on the demo program for the C generation schemes.
 * It shows some basic generation tasks.
 */

#include <cstddef>
#include <cstdint>
#include <cstdlib>
#include <cassert>
#include <climits>
#include <iostream>
#include <sstream>
#include <iomanip>
#include <algorithm>
#include <numeric>
#include <random>       // for random_device

#include "pcg_random.hpp"

// This code can be compiled with the preprocessor symbol RNG set to the
// PCG generator you'd like to test.

#ifndef RNG
    #define RNG pcg32
    #define TWO_ARG_INIT 1
#endif

#define STRINGIFY_IMPL(x) #x
#define STRINGIFY(x)      STRINGIFY_IMPL(x)

using namespace std;
using pcg_extras::operator<<;

#if !PCG_EMULATED_128BIT_MATH || !AWKWARD_128BIT_CODE

int main(int argc, char** argv)
{
    // Read command-line options

    int rounds = 5;
    bool nondeterministic_seed = false;

    ++argv;
    --argc;
    if (argc > 0 && strcmp(argv[0], "-r") == 0) {
        nondeterministic_seed = true;
        ++argv;
        --argc;
    }
    if (argc > 0) {
        rounds = atoi(argv[0]);
    }

    /* Many of the generators can be initialized with two arguments; the second
     * one specifies the stream.
     */

#if TWO_ARG_INIT
    RNG rng(42u, 54u);
#else
    RNG rng(42u);
#endif

    if (nondeterministic_seed) {
        // Seed with external entropy from std::random_device (a handy
        // utility provided by pcg_extras).
        rng.seed(pcg_extras::seed_seq_from<std::random_device>());
    }

    constexpr auto bits = sizeof(RNG::result_type) * CHAR_BIT;
    constexpr int how_many_nums = bits <= 8  ? 14
                                : bits <= 16 ? 10
                                :              6;
    constexpr int wrap_nums_at  = bits >  64 ? 2
                                : bits >  32 ? 3
                                :              how_many_nums;

    cout << STRINGIFY(RNG) << ":\n"
    //   << "      -  aka:         " << pcg_extras::printable_typename<RNG>()
    // ^-- we skip this line because the output is long, scary, ugly, and
    //     and varies depending on the platform
         << "      -  result:      " <<  bits << "-bit unsigned int\n"
         << "      -  period:      2^" << RNG::period_pow2();
    if (RNG::streams_pow2() > 0)
         cout << "   (* 2^" << RNG::streams_pow2() << " streams)";
    cout << "\n      -  size:        " << sizeof(RNG) << " bytes\n\n";

    for (int round = 1; round <= rounds; ++round) {
        printf("Round %d:\n", round);

        /* Make some N-bit numbers */
        cout << setw(4) << setfill(' ') << bits << "bit:";
        for (int i = 0; i < how_many_nums; ++i) {
            if (i > 0 && i % wrap_nums_at == 0)
                cout << "\n\t";
            cout << " 0x" << hex << setfill('0') 
                 << setw(sizeof(RNG::result_type)*2) << rng();
        }
        cout << endl;

        cout << "  Again:";
        rng.backstep(6);
        for (int i = 0; i < how_many_nums; ++i) {
            if (i > 0 && i % wrap_nums_at == 0)
                cout << "\n\t";
            cout << " 0x" << hex << setfill('0') 
                 << setw(sizeof(RNG::result_type)*2) << rng();
        }
        cout << dec << endl;

        /* Toss some coins */
        cout << "  Coins: ";
        for (int i = 0; i < 65; ++i)
            cout << (rng(2) ? "H" : "T");
        cout << endl;

        RNG rng_copy{rng};
        /* Roll some dice */
        printf("  Rolls:");
        for (int i = 0; i < 33; ++i)
            cout << " " << (uint32_t(rng(6)) + 1);
        cout << "\n   -->   rolling dice used "
             << static_cast<size_t>(rng - rng_copy)
             << " random numbers" << endl;

        /* Deal some cards using pcg_extras::shuffle, which follows
         * the algorithm for shuffling that most programmers would expect.
         * (It's unspecified how std::shuffle works.)
         */      
        enum { SUITS = 4, NUMBERS = 13, CARDS = 52 };
        char cards[CARDS];
        iota(begin(cards), end(cards), 0);
        pcg_extras::shuffle(begin(cards), end(cards), rng);

        /* Output the shuffled deck */
        printf("  Cards:");
        static const signed char number[] = {'A', '2', '3', '4', '5', '6', '7',
                                             '8', '9', 'T', 'J', 'Q', 'K'};
        static const signed char suit[] =   {'h', 'c', 'd', 's'};
        int i = 0;
        for (auto card : cards) {
            ++i;
            cout << " " << number[card / SUITS] << suit[card % SUITS];
            if (i % 22 == 0)
                cout << "\n\t";
        }
        
        cout << "\n" << endl;
    }

    // Ensure that the input/output functions work.
    std::stringstream ss;
    ss << rng;
    RNG from_string;
    ss >> from_string;
    assert(from_string == rng);

    return 0;
}

#else //  i.e. PCG_EMULATED_128BIT_MATH && AWKWARD_128BIT_CODE

int main()
{
    // It's not that it *can't* be done, it just requires either C++14-style
    // constexpr or some things not to be declared const.
    cout << "Sorry, " STRINGIFY(RNG) " not supported with emulated 128-bit math"
         << endl;
}

#endif
