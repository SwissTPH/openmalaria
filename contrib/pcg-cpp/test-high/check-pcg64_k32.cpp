#define RNG pcg64_k32
#define TWO_ARG_INIT 1
#define AWKWARD_128BIT_CODE 1

#include "pcg-test.cpp"

