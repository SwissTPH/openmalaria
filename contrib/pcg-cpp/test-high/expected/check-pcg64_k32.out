pcg64_k32:
      -  result:      64-bit unsigned int
      -  period:      2^2176   (* 2^127 streams)
      -  size:        288 bytes

Round 1:
  64bit: 0x2dcbaf9339a0a8db 0xa5486595bb7ecc26 0x800934faba0d0759
	 0x7eff732dee5b72cb 0x44760d3bb06156f9 0x6f3622730aff6c44
  Again: 0x2dcbaf9339a0a8db 0xa5486595bb7ecc26 0x800934faba0d0759
	 0x7eff732dee5b72cb 0x44760d3bb06156f9 0x6f3622730aff6c44
  Coins: THTTHHHHHHTHTTHHTHTTTHTTTTTHTTHHHHHHHHHHHHTTHHHHTTHTHTTHTHTHTHHHH
  Rolls: 4 3 1 4 1 6 1 3 6 2 1 4 2 1 5 4 5 6 5 4 6 1 5 5 1 2 2 2 2 6 2 6 4
   -->   rolling dice used 33 random numbers
  Cards: 2s 2d 9s Jd 3c 4h Qs 5s 4c 2h 9c 6s Qd Kh Ac Ah 7h 8h 6c Ks 5c Tc
	 Qh 8d 8s Kd 4s Th 9h 3h Ts 7s Td 5d 2c 4d Kc Qc 3d As 6h 8c 6d 7c
	 Jh Ad 3s 5h 9d 7d Js Jc

Round 2:
  64bit: 0x08362f38c5184d06 0x2ba1b3164afc696d 0xf18e00ded413293e
	 0x14512048d47657aa 0x003ef118b96e494b 0xb6ba214a64598dc5
  Again: 0x08362f38c5184d06 0x2ba1b3164afc696d 0xf18e00ded413293e
	 0x14512048d47657aa 0x003ef118b96e494b 0xb6ba214a64598dc5
  Coins: TTHHHHTHTTHTTHTTTTHHHHHHTTTHTHHHHHTTTTTHTTTTHTHHHTHTHTTHTTTTTHHTT
  Rolls: 1 6 4 6 6 6 6 4 1 4 5 6 1 4 4 6 1 3 5 2 3 2 3 5 2 4 5 2 6 6 3 1 1
   -->   rolling dice used 33 random numbers
  Cards: Th Jd Tc 7d Qd 3d 2c Qs 8c 6d 2h 7s Jh 9d Td 6c Ac 8s 7c 8h 9s Jc
	 4d Kc 2s 7h 2d 5c 9c 9h Ad Kh 4c Qh 3h 3c 5d 3s 5s Qc 5h Ts 8d Js
	 Ah Ks As 4h Kd 4s 6h 6s

Round 3:
  64bit: 0xedfa63d9e74cb0d7 0xb9a431ed222e20e3 0xe0bc6a0b891ffa90
	 0x8e1c3e7e25a97e13 0x7065328d5502585a 0x30ea18f29c9dd917
  Again: 0xedfa63d9e74cb0d7 0xb9a431ed222e20e3 0xe0bc6a0b891ffa90
	 0x8e1c3e7e25a97e13 0x7065328d5502585a 0x30ea18f29c9dd917
  Coins: HTHHTTHHTTTHHHHTTHTTTHHTHTHHTHHHTTHTTTHHHHTHHTHTHHTHHTTHTHHHTTHHH
  Rolls: 5 6 2 6 2 1 2 1 5 4 6 4 6 1 4 5 6 6 4 5 3 6 1 6 4 4 3 3 3 5 6 6 6
   -->   rolling dice used 33 random numbers
  Cards: Th 8d Jh 2s Td 3h 2d Kc 3c 7d 6s Jc 5c Jd 8c Qd 6d As 4s 5d Kh 6h
	 Js 7h 8s 9h Ad Qs 2c 9s 4d 5h 8h 4c 6c 3d 9d 3s Ac Tc 5s Qh Qc 7c
	 2h 9c Kd 7s 4h Ah Ks Ts

Round 4:
  64bit: 0x778fbeeb1240be87 0x146e38cf2b84691c 0x51f4076de124d878
	 0x3aa6741428102d8b 0xfe712690a64ff70b 0xfdbb7799320fb4a3
  Again: 0x778fbeeb1240be87 0x146e38cf2b84691c 0x51f4076de124d878
	 0x3aa6741428102d8b 0xfe712690a64ff70b 0xfdbb7799320fb4a3
  Coins: THTTTHTTHHHTHHTTTTTHTTTHHTTHHHTTTTTHTHTHHTHTHHTTHTTTTHTHHTTHTTHHH
  Rolls: 6 1 6 6 1 2 2 1 3 3 1 5 6 3 3 2 6 1 3 2 3 4 6 6 6 2 3 4 6 1 3 5 1
   -->   rolling dice used 33 random numbers
  Cards: 7s Kd 3d Tc 9s Ah 6d 8d 3s Ad Kc 6s 9h Kh 4c Qs Ks Js Th 3h 5s 4h
	 Ts 5c 7d 4d 7c 9c 5d 8h 2s 2h 6h 6c 4s Jc Jd As Td Qd Qh 8s 2d Qc
	 7h 5h Ac 3c 8c 9d Jh 2c

Round 5:
  64bit: 0xc698f9ba129692a7 0x94646b27c1c8ca84 0x5ed4146e49289180
	 0x646e2e367a11b087 0xcf0ec8f5395d2d9f 0xb81f13d75e8265af
  Again: 0xc698f9ba129692a7 0x94646b27c1c8ca84 0x5ed4146e49289180
	 0x646e2e367a11b087 0xcf0ec8f5395d2d9f 0xb81f13d75e8265af
  Coins: THTHTTHHTHTTHHHHTHTTTTTHTHTHTHTTTTHHHTHTTTTHTHHTTHTHTTHHHTHTTTHTT
  Rolls: 6 2 1 3 2 3 3 5 1 1 4 5 2 2 4 2 5 6 1 3 2 4 4 1 2 4 5 4 5 2 3 6 1
   -->   rolling dice used 33 random numbers
  Cards: Ac 3d 8s 4c 5h 7c 4d 9c 7d Kc 4h Qc Td 8c Js 9s Tc Jd Qs 3h Jh 6d
	 2h Ks 5s Ts 7s 5d 2c 2d 2s 5c 7h As 3s Qh Kh 9d Th 9h 8h Jc Ah Qd
	 Kd 8d 6c 4s 3c 6s Ad 6h

